### PR TITLE
OAuth 2.0 get token work item handler implementation

### PR DESCRIPTION
**Pull request by marianbuenosayres**
This PR has an implementation to request a token from an OAuth 2.0 source

### CODE REVIEW

File: core-wih/src/main/java/io/elimu/a2d2/corewih/OAuth2WorkItemHandler.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	@Override
+	public void executeWorkItem(WorkItem workItem, WorkItemManager manager) {
+		String clientId = (String) workItem.getParameter("clientId");
+		String scope = (String) workItem.getParameter("scope");
+		String username = (String) workItem.getParameter("username");
+		String password = (String) workItem.getParameter("password");
+		String grantType = (String) workItem.getParameter("grantType");
+		String tokenUrl = (String) workItem.getParameter("tokenUrl");
+		List<String> missingValues = new ArrayList<>();
+		if (clientId == null || "".equals(clientId.trim())) {
```
Comment (hankwallace) [inline core-wih/src/main/java/io/elimu/a2d2/corewih/OAuth2WorkItemHandler.java:42]: Consider using `StringUtils.isEmpty(str)` from Apache Commons to clean up these checks.

File: core-wih/src/main/java/io/elimu/a2d2/corewih/OAuth2WorkItemHandler.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	        os.print(content);
+	        os.close();
+	        reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
+	        String line = null;
+	        StringWriter out = new StringWriter(connection.getContentLength() > 0 ? connection.getContentLength() : 2048);
+	        while ((line = reader.readLine()) != null) {
+	            out.append(line);
+	        }
+	        String response = out.toString();
+	        Matcher matcher = patToken.matcher(response);
```
Comment (hankwallace) [inline core-wih/src/main/java/io/elimu/a2d2/corewih/OAuth2WorkItemHandler.java:89]: Why are you using a pattern matcher instead of a Json parser here?

**Review by hankwallace — COMMENTED** [review]
Finished with my first pass. I'll review again after the comments in the design document are discussed.

File: core-wih/src/main/java/io/elimu/a2d2/corewih/OAuth2WorkItemHandler.java (hunk 1 above)
Comment (marianbuenosayres) [reply core-wih/src/main/java/io/elimu/a2d2/corewih/OAuth2WorkItemHandler.java]: Changed to StringUtils.isEmpty

File: core-wih/src/main/java/io/elimu/a2d2/corewih/OAuth2WorkItemHandler.java (hunk 2 above)
Comment (marianbuenosayres) [reply core-wih/src/main/java/io/elimu/a2d2/corewih/OAuth2WorkItemHandler.java]: Changed to use GSON. It seemed simpler when I was just extracting the access_token, but the more I added as an output the more it made sense to use a parser :)

File: core-wih/src/main/java/io/elimu/a2d2/corewih/OAuth2WorkItemHandler.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	        os.close();
+	        reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
+	        String line = null;
+	        StringWriter out = new StringWriter(connection.getContentLength() > 0 ? connection.getContentLength() : 2048);
+	        while ((line = reader.readLine()) != null) {
+	            out.append(line);
+	        }
+	        String response = out.toString();
+	        Gson gson = new Gson();
+	        Map<?, ?> output = (Map<?, ?>) gson.fromJson(response, Map.class);
```
Comment (hankwallace) [inline core-wih/src/main/java/io/elimu/a2d2/corewih/OAuth2WorkItemHandler.java:87]: The following fields are also in the result, but I'm not confident that we need them.
`token_type`, `session_state`, and `scope`

File: core-wih/src/main/java/io/elimu/a2d2/corewih/OAuth2WorkItemHandler.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	        }
+	    } catch (Exception e) {
+	    	results.put("error", e);
+	    	results.put("errorMessage", e.getMessage());
+	    } finally {
+	        if (reader != null) {
+	            try {
+	                reader.close();
+	            } catch (IOException e) {
+	            }
```
Comment (hankwallace) [inline core-wih/src/main/java/io/elimu/a2d2/corewih/OAuth2WorkItemHandler.java:108]: Should we log something here?

**Review by hankwallace — DISMISSED** [review]
A couple of questions, but looks pretty good.

File: core-wih/src/main/java/io/elimu/a2d2/corewih/OAuth2WorkItemHandler.java (hunk 3 above)
Comment (marianbuenosayres) [reply core-wih/src/main/java/io/elimu/a2d2/corewih/OAuth2WorkItemHandler.java]: Added them as outputs

File: cds-hook-model/src/main/java/io/elimu/a2d2/oauth/OAuthUtils.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+	        }
+	        connection.disconnect();
+	    }
+	    return results;
+	}
+
+	private static void longify(Map<String, Object> results, String key) {
+		if (results.containsKey(key)) {
+			try {
+				String value = String.valueOf(results.get("expires_in")).replace(".0", "");
```
Comment (kevinliuelimu) [inline cds-hook-model/src/main/java/io/elimu/a2d2/oauth/OAuthUtils.java:94]: not sure the purpose of  "expires_in"  @ln94? Is this function only for key "expires_in" and "refresh_expires_in"
also,  replace(".0", "") may not be correct here.  Yes, "100.0" becomes 100, but what if the value is "100.00", that would be 1000.

File: cds-hook-model/src/main/java/io/elimu/a2d2/oauth/OAuthUtils.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+	        String response = out.toString();
+	        results = new Gson().fromJson(response, new TypeToken<Map<String, Object>>() {
+				private static final long serialVersionUID = 1L; 
+	        }.getType());
+	        longify(results, "expires_in");
+	        longify(results, "refresh_expires_in");
+	    } catch (Exception e) {
+	    	results.put("error", e);
+	    	results.put("errorMessage", e.getMessage());
+	    } finally {
```
Comment (kevinliuelimu) [inline cds-hook-model/src/main/java/io/elimu/a2d2/oauth/OAuthUtils.java:78]: can we use try-with-resources here instead?

**Review by kevinliuelimu — COMMENTED** [review]
Please add copyright headers to new files.  I also have some comments inline

File: cds-hook-model/src/main/java/io/elimu/a2d2/oauth/OAuthUtils.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+import com.google.gson.Gson;
+
+public class OAuthUtils {
+	
+	private static final Logger LOG = LoggerFactory.getLogger(OAuthUtils.class);
+	private static final Map<String, Map<String, Object>> CACHE = new HashMap<>();
+
+	//TODO invoke this from QueryingServerHelper
+	public static Map<String, Object> authenticate(String body, String tokenUrl, String clientId, String clientSecret) {
+		String keyUnhashed = "url=" + tokenUrl + "&" + body + "&client_id=" + clientId + "&client_secret=" + clientSecret;
```
Comment (hankwallace) [inline cds-hook-model/src/main/java/io/elimu/a2d2/oauth/OAuthUtils.java:29]: We may need more than `client_secret` here. In cases where there is no client secret, but there is a `username` and `password`, do we want the cache to serve up the same token for different credentials?

File: cds-hook-model/src/main/java/io/elimu/a2d2/oauth/OAuthUtils.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+		if (retval.containsKey("expires_in")) {
+			Long expirationTime = ((Long) retval.get("expires_in")).longValue() * 1000 + System.currentTimeMillis();
+			retval.put("_timeutil_expiration_time_millis", expirationTime);
+			CACHE.put(key, retval);
+		}
+		return retval;
+	}
+
+	private static Map<String, Object> _authenticate(String body, String tokenUrl, String clientId, String clientSecret) {
+		BufferedReader reader = null;
```
Comment (hankwallace) [inline cds-hook-model/src/main/java/io/elimu/a2d2/oauth/OAuthUtils.java:48]: The indentation here looks wrong.

File: cds-hook-model/src/main/java/io/elimu/a2d2/oauth/OAuthUtils.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+		if (results.containsKey(key)) {
+			try {
+				String value = String.valueOf(results.get("expires_in")).replace(".0", "");
+				results.put(key, Long.valueOf(value));
+			} catch (Exception e) {
+				LOG.warn("Couldn't parse value '" + results.get(key) + "' from key '" + key + "' as long");
+			}
+		}
+	}
+}
```
Comment (hankwallace) [inline cds-hook-model/src/main/java/io/elimu/a2d2/oauth/OAuthUtils.java:101]: It's a nit, but files should have an ending newline.

File: core-wih/src/main/java/io/elimu/a2d2/corewih/OAuth2WorkItemHandler.java [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+		if (StringUtils.isEmpty(scope)) {
+			scope = "offline_access";
+		}
+		if (StringUtils.isEmpty(clientId)) {
+			missingValues.add("clientId");
+		}
+		if (StringUtils.isEmpty(grantType)) {
+			missingValues.add("grantType");
+		} else {
+			switch (grantType) {
```
Comment (hankwallace) [inline core-wih/src/main/java/io/elimu/a2d2/corewih/OAuth2WorkItemHandler.java:47]: 👍

**Review by hankwallace — CHANGES_REQUESTED** [review]
Finished with my review. A couple of minor suggestions in addition to Kevin's.

File: cds-hook-model/src/main/java/io/elimu/a2d2/oauth/OAuthUtils.java (hunk 7 above)
Comment (marianbuenosayres) [reply cds-hook-model/src/main/java/io/elimu/a2d2/oauth/OAuthUtils.java:29]: @hankwallace in those cases, the content will be coming from the body variable, which is built with BodyBuilder using all parameters available for the authentication

File: cds-hook-model/src/main/java/io/elimu/a2d2/oauth/OAuthUtils.java (hunk 6 above)
Comment (marianbuenosayres) [reply cds-hook-model/src/main/java/io/elimu/a2d2/oauth/OAuthUtils.java:78]: I added it for the reader but the connection is not a closeable resource and instead needs to use "disconnect()"

File: cds-hook-model/src/main/java/io/elimu/a2d2/oauth/OAuthUtils.java (hunk 9 above)
Comment (marianbuenosayres) [reply cds-hook-model/src/main/java/io/elimu/a2d2/oauth/OAuthUtils.java]: Added

File: cds-hook-model/src/main/java/io/elimu/a2d2/oauth/OAuthUtils.java (hunk 5 above)
Comment (marianbuenosayres) [reply cds-hook-model/src/main/java/io/elimu/a2d2/oauth/OAuthUtils.java]: Sorry, that was my mistake from a late night refactor. Corrected it now to work properly

**Comment by marianbuenosayres** [comment]
> Finished with my review. A couple of minor suggestions in addition to Kevin's.

This is marked as a change request but has no specific changes associated to this comment. Can you resolve or remove it on your next review @hankwallace ?

**Review by hankwallace — APPROVED** [review]

**Pull request merged**